src/modes/run.ts

```diff
@@ -1,5 +1,340 @@
 
-export async function run(_options: RunOptions) {
-  throw new Error("Not implemented");
+/**
+ * Run the watch-tower 👀🐮
+ * @param options Specified by the CLI / environment for running the watch-tower
+ */
+export async function run(options: RunOptions) {
+  const { rpc, deploymentBlock, oneShot } = options;
+
+  process.on("unhandledRejection", async (error) => {
+    console.log(error);
+    await DBService.getInstance().close();
+    process.exit(1);
+  });
+
+  process.on("SIGINT", async function () {
+    console.log("️⚠️ Caught interrupt signal ⚠️");
+    await DBService.getInstance().close();
+    process.exit(0);
+  });
+
+  try {
+    const chainWatchers = await Promise.all(
+      rpc.map((rpc, index) => {
+        return ChainWatcher.init(
+          {
+            ...options,
+            rpc,
+            deploymentBlock: deploymentBlock[index],
+          },
+          DBService.getInstance()
+        );
+      })
+    );
+
+    // Run the block watcher for each chain
+    const runPromises = chainWatchers.map(async (chainWatcher) => {
+      return chainWatcher.warmUp(oneShot);
+    });
+
+    // Run all the chain watchers
+    await Promise.all(runPromises);
+  } catch (error) {
+    console.error(error);
+    await DBService.getInstance().close();
+    process.exit(1);
+  } finally {
+    await DBService.getInstance().close();
+  }
+}
+
+/**
+ * The chain watcher handles watching a single chain for new conditional orders
+ * and executing them.
+ */
+export class ChainWatcher {
+  private readonly deploymentBlock: number;
+  private readonly pageSize: number;
+  private readonly publish: boolean;
+  private inSync = false;
+
+  chainContext: ChainContext;
+  registry: Registry;
+
+  protected constructor(
+    options: SingularRunOptions,
+    chainContext: ChainContext,
+    registry: Registry
+  ) {
+    const { deploymentBlock, pageSize, publish } = options;
+    this.deploymentBlock = deploymentBlock;
+    this.pageSize = pageSize;
+    this.publish = publish;
+
+    this.registry = registry;
+    this.chainContext = chainContext;
+  }
+
+  public static async init(
+    options: SingularRunOptions,
+    storage: DBService
+  ): Promise<ChainWatcher> {
+    const { rpc, deploymentBlock } = options;
+
+    const chainContext = await ChainContext.create(rpc);
+    const registry = await Registry.load(
+      storage,
+      chainContext.chainId.toString(),
+      deploymentBlock
+    );
+
+    return new ChainWatcher(options, chainContext, registry);
+  }
+
+  /**
+   * Warm up the chain watcher by fetching the latest block number and
+   * checking if the chain is in sync.
+   * @returns the run promises for what needs to be watched
+   */
+  public async warmUp(oneShot?: boolean) {
+    const { provider, chainId } = this.chainContext;
+    const { lastProcessedBlock } = this.registry;
+    const { pageSize } = this;
+    const _ = (s: string) => console.log(`[warmUp:chainId:${chainId}] ${s}`);
```

Comment:
Please no variables named `_` 😆 

If you intend to keep it, make it a private method that returns the log fn for example.

```ts
private getLog(chainId:number): (s:string)=>void {
  return (s:string) => console.log(`[warmUp:chainId:${chainId}] ${s}`)
}
```

Reply:
Unified logging tracked in #60